### PR TITLE
CollectQualityYieldMetrics in Spark and tests for it

### LINKED ISSUE #999: CollectQualityYieldMetrics in Spark
the requirement is to implement CollectQualityYieldMetrics in gatk using spark and have the output match the picard output.

### PR DESCRIPTION
**Pull request by akiezun**
fixes #999 
for @davidadamsphd

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        //NOTE: this class must have a 0-arg constructor so we need to use a setter for this field.
+        //Also, we want to be able to easily construct usable instances so the setter returns 'this' so we can say
+        //new QualityYieldMetrics().setUseOriginalQualities(true) in the aggregate call
+        public QualityYieldMetrics setUseOriginalQualities(final boolean useOriginalQualities){
+            this.useOriginalQualities = useOriginalQualities;
+            return this;
+        }
+
+        /** The total number of reads in the input file */
+        public int TOTAL_READS;
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: These variables don't match Google/Hellbender style (should be `totalReads`).

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+        /** The number of reads that are PF - pass filter */
+        public int PF_READS;
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: Do these really need to be public?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        /** The total number of bases in all PF reads */
+        public long PF_BASES;
+
+        /** The number of bases in all reads that achieve quality score 20 or higher */
+        public long Q20_BASES;
+
+        /** The number of bases in PF reads that achieve quality score 20 or higher */
+        public long PF_Q20_BASES;
+
+        /** The number of bases in all reads that achieve quality score 20 or higher */
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: *30* or higher

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        public long Q30_BASES;
+
+        /** The number of bases in PF reads that achieve quality score 20 or higher */
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: *30* or higher.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+                if (tmp == null) tmp = read.getBaseQualities();
+                quals = tmp;
+            } else {
+                quals = read.getBaseQualities();
+            }
+
+            // add up quals, and quals >= 20
+            for (byte qual : quals) {
+                Q20_EQUIVALENT_YIELD += qual;
+                if (qual >= 20) Q20_BASES++;
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: nit: here and elsewhere, put braces around if's.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                    "as inputs instead of the quality scores in the QUAL field.")
+    public boolean useOriginalQualities = false;
+
+
+    /** A set of metrics used to describe the general quality of a BAM file */
+    public static final class QualityYieldMetrics extends MetricBase implements Serializable{
+        private static final long serialVersionUID = 1;
+
+        private boolean useOriginalQualities;
+        private boolean done;
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: I think finalized as the variable makes the use clearer.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+    private MetricsFile<QualityYieldMetrics,Integer> makeMetricsFile(QualityYieldMetrics metrics) {
+        final MetricsFile<QualityYieldMetrics, Integer> metricsFile = getMetricsFile();
+        metricsFile.addMetric(metrics);
+        return metricsFile;
+    }
+
+    private void saveResults(final MetricsFile<QualityYieldMetrics, Integer> metricsFile) {
+        final GCSOptions gcsOptions = getAuthenticatedGCSOptions(); // null if we have no api key
+        try(OutputStream outputStream = BucketUtils.createFile(out, gcsOptions)) {
```
Comment (davidadamsphd) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: I filed a issues to move this into a library (#1025, #1026).

**Comment by davidadamsphd** [comment]
As with the last few, nothing serious, just small comments.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: I know. The problem is that those metrics are printed out to this file that is consumed by some Broad tools downstream and those expect this format. I'll add comments about this throughout

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: unfortunately yes. The superclass only finds them if they are public and it needs to find them to print them to the metrics file.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: nice catch, I fixed it in the non-spark version too

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java (hunk 4 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: nice catch, I fixed it in the non-spark version too

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: yup. done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: ok

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectQualityYieldMetricsSpark.java]: thanks

**Comment by akiezun** [comment]
back to you @davidadamsphd

**Comment by davidadamsphd** [comment]
LGTM.

**Comment by davidadamsphd** [comment]
(Squash, then submit).

**Pull request merged**